### ISSUE TITLE
Analyzer should warn on identifiers named underscore

### ISSUE THREAD

**Issue by jcouv** [dotnet/roslyn#26594]
Like `var` and `nameof`, the compiler gives two meanings to `_` depending if one was already declared. This can be ambiguous.
Such an analyzer would help customers avoid the ambiguity and only use `_` for discards.

This should probably be turned off by default to avoid compat issue.

**Comment by DavidArno** [comment]
Could that analyser warn on using `var`, `nameof` etc too, or is that broadening the scope too much?

**Comment by CyrusNajmabadi** [comment]
> This should probably be turned off by default to avoid compat issue.

I disagree.  It should probably be on, but as a suggestion.  

I also agree with david, this should help encourage people to move away from 'var' or 'nameof'.  If someone does not want that, they can def just suppress, or just disable the analyzer.

**Comment by sharwell** [comment]
More specifically, the analyzer should warn on the *use* of an identifier named `_`.

Also the analyzer could warn on the following:

```
// diagnostic
var _ = expression;

// use this instead:
_ = expression;
```

**Comment by OJacot-Descombes** [comment]
... and syntax coloring should color it as keyword as it does already for `var` and `nameof`.

**Comment by Joe4evr** [comment]
> ... and syntax coloring should color it as keyword as it does already for `var` and `nameof`.

Back in 2017 previews, use of `_` as a discard *was* colored as a keyword. Why was that removed in the first place?

**Comment by sharwell** [comment]
We still color `_` as a discard when it is used as a discard. We color it as an identifier when it is used as an identifier.

**Comment by Joe4evr** [comment]
@sharwell That's not what it looks like to me:
![DiscardDoubts](https://user-images.githubusercontent.com/1058562/68633557-9795e000-04f2-11ea-981f-3a947ae22611.png)

**Comment by sharwell** [comment]
Nice, it works for `out var _` but doesn't work for `out _`. That's a bug.

**Comment by sharwell** [comment]
@Joe4evr I filed #39768 for the case you mentioned
